overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,270 @@
+	st.Set("snapshot-set", setState)
+
+	return setState.LastID, nil
+}
+
+func allActiveSnapNames(st *state.State) ([]string, error) {
+	all, err := snapstateAll(st)
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(all))
+	for name, snapst := range all {
+		if snapst.IsInstalled() {
+			names = append(names, name)
+		}
+	}
+
+	sort.Strings(names)
+
+	return names, nil
+}
+
+func snapNamesInSnapshotSet(setID uint64, requested []string) (snapsFound []string, filenames []string, err error) {
+	sort.Strings(requested)
+	found := false
+	err = backendIter(context.Background(), func(r *backend.Reader) error {
+		if r.SetID == setID {
+			found = true
+			if len(requested) == 0 || strutil.SortedListContains(requested, r.Snap) {
+				snapsFound = append(snapsFound, r.Snap)
+				filenames = append(filenames, r.Name())
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if !found {
+		return nil, nil, client.ErrSnapshotSetNotFound
+	}
+	if len(snapsFound) == 0 {
+		return nil, nil, client.ErrSnapshotSnapsNotFound
+	}
+
+	return snapsFound, filenames, nil
+}
+
+// checkSnapshotTaskConflict checks whether there's an in-progress task for snapshots with the given set id.
+func checkSnapshotTaskConflict(st *state.State, setID uint64, conflictingKinds ...string) error {
+	for _, task := range st.Tasks() {
+		if task.Change().Status().Ready() {
+			continue
+		}
+		if !strutil.ListContains(conflictingKinds, task.Kind()) {
+			continue
+		}
+
+		var snapshot snapshotState
+		if err := task.Get("snapshot", &snapshot); err != nil {
+			if err == state.ErrNoState {
+				return fmt.Errorf("internal error: task %s (%s) is missing snapshot information", task.ID(), task.Kind())
+			}
+			return err
+		}
+
+		if snapshot.SetID == setID {
+			return fmt.Errorf("snapshot set #%d has a %q task in progress", setID, task.Kind())
+		}
+	}
+
+	return nil
+}
+
+// List valid snapshots.
+// Note that the state must be locked by the caller.
+var List = backend.List
+
+// Save creates a taskset for taking snapshots of snaps' data.
+// Note that the state must be locked by the caller.
+func Save(st *state.State, snapNames []string, users []string) (setID uint64, snapsSaved []string, ts *state.TaskSet, err error) {
+	if len(snapNames) == 0 {
+		snapNames, err = allActiveSnapNames(st)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+	}
+
+	// Make sure we do not snapshot if anythink like install/remove/refresh is in progress
+	if err := snapstateCheckChangeConflictMany(st, snapNames, nil); err != nil {
+		return 0, nil, nil, err
+	}
+
+	setID, err = newSnapshotSetID(st)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	ts = state.NewTaskSet()
+
+	for _, name := range snapNames {
+		desc := fmt.Sprintf("Save the data of snap %q in snapshot set #%d", name, setID)
+		task := st.NewTask("save-snapshot", desc)
+		snapshot := snapshotState{
+			SetID: setID,
+			Snap:  name,
+			Users: users,
+		}
+		task.Set("snapshot", &snapshot)
```

Comment:
the type and name are strange here,  convention would be more "snapshot-setup" and snapshotSetup for the type.

Or is this also stored permanently in the state?

Reply:
you're right (and so I went with snapshot-setup)